cli/cmd/cluster/certificate-rotator.go

```diff
@@ -0,0 +1,227 @@
+package cluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
+	"github.com/kinvolk/lokomotive/pkg/platform"
+)
+
+type certificateRotatorConfig struct {
+	clientSet            *kubernetes.Clientset
+	newCACert            string
+	logger               *log.Entry
+	daemonSetsToRestart  []platform.Workload
+	deploymentsToRestart []platform.Workload
+}
+
+type certificateRotator struct {
+	config certificateRotatorConfig
+}
+
+// CertificateRotateOptions contains the options for the RotateCertificates function.
+type CertificateRotateOptions struct {
+	Confirm    bool
+	Verbose    bool
+	ConfigPath string
+	ValuesPath string
+}
+
+// RotateCertificates replaces all certificates in a cluster.
+// due to the nature of it running as a lokoctl command it should be idempotency.
+func RotateCertificates(contextLogger *log.Entry, options CertificateRotateOptions) error {
+	cc := clusterConfig{
+		verbose:    options.Verbose,
+		configPath: options.ConfigPath,
+		valuesPath: options.ValuesPath,
+	}
+
+	c, err := cc.initialize(contextLogger)
+	if err != nil {
+		return fmt.Errorf("initializing: %w", err)
+	}
+
+	if err := canRotate(c); err != nil {
+		return fmt.Errorf("cannot rotate cluster: %w", err)
+	}
+
+	// Tainting certificates so they get rotated.
+	if err := c.taintCertificates(); err != nil {
+		rollbackErr := c.untaintCertificates()
+		if rollbackErr != nil {
+			return fmt.Errorf("tainting certificate resources, rolled back successfully: %w", err)
+		}
+
+		return fmt.Errorf("tainting certificate resources: %w, rolled back gave error: %v", err, rollbackErr)
+	}
+
+	// Apply the Terraform changes to replace tainted resources.
+	if err := c.platform.ApplyWithoutParallel(&c.terraformExecutor); err != nil {
```

Comment:
Why without parallelism?

UPDATE: Ohh, I guess to not potentially break all etcd at the same time? Can we add a comment for that? I know I'll try to remove it when touching the code if I don't see why :-D